File: server/services/s3_service.py
import os
import boto3
import uuid
import logging
from config import Config

logger = logging.getLogger(__name__)

def get_s3_client():
    if not Config.AWS_ACCESS_KEY_ID or not Config.AWS_SECRET_ACCESS_KEY:
        logger.warning("AWS Credentials not configured in environment.")
        return None
    return boto3.client(
        "s3",
        aws_access_key_id=Config.AWS_ACCESS_KEY_ID,
        aws_secret_access_key=Config.AWS_SECRET_ACCESS_KEY,
        region_name=Config.AWS_REGION
    )

def upload_file_to_s3(file_data, file_name, content_type):
    """
    Uploads a file stream/bytes to AWS S3 bucket and returns the public URL.
    """
    s3_client = get_s3_client()
    if not s3_client:
        raise ValueError(
            "AWS S3 is not configured. Please add AWS_ACCESS_KEY_ID, "
            "AWS_SECRET_ACCESS_KEY, and AWS_S3_BUCKET to your backend .env file."
        )

    if not Config.AWS_S3_BUCKET:
        raise ValueError("AWS_S3_BUCKET is not set in the environment variables.")

    # Generate a unique key
    ext = os.path.splitext(file_name)[1]
    unique_key = f"avatars/{uuid.uuid4().hex}{ext}"

    try:
        s3_client.put_object(
            Bucket=Config.AWS_S3_BUCKET,
            Key=unique_key,
            Body=file_data,
            ContentType=content_type
        )
        # Construct standard public S3 URL
        url = f"https://{Config.AWS_S3_BUCKET}.s3.{Config.AWS_REGION}.amazonaws.com/{unique_key}"
        logger.info(f"File uploaded successfully to S3: {url}")
        return url
    except Exception as e:
        logger.exception("S3 upload failed:")
        raise RuntimeError(f"AWS S3 upload failed: {str(e)}")

def delete_file_from_s3(image_url):
    """
    Parses an S3 object URL and deletes it from the S3 bucket to prevent orphan files.
    """
    if not image_url:
        return False

    s3_client = get_s3_client()
    if not s3_client:
        logger.warning("AWS credentials not configured. Skipping S3 deletion.")
        return False

    if not Config.AWS_S3_BUCKET:
        logger.warning("AWS_S3_BUCKET not configured. Skipping S3 deletion.")
        return False

    try:
        # Verify if it looks like an AWS S3 URL
        if "s3" not in image_url or ".amazonaws.com" not in image_url:
            return False

        # Extract the S3 Key from the URL
        # e.g., https://my-bucket.s3.us-east-1.amazonaws.com/avatars/abc123xyz.png
        if "avatars/" in image_url:
            key = "avatars/" + image_url.split("avatars/")[1]
        else:
            # Fallback parsing
            parts = image_url.split(".amazonaws.com/")
            if len(parts) > 1:
                key = parts[1]
            else:
                return False

        s3_client.delete_object(Bucket=Config.AWS_S3_BUCKET, Key=key)
        logger.info(f"Successfully deleted S3 object: {key}")
        return True
    except Exception as e:
        logger.exception(f"Failed to delete S3 object for url {image_url}:")
        return False

File: backend/src/services/s3Service.js
import { S3Client, PutObjectCommand, DeleteObjectCommand } from "@aws-sdk/client-s3";
import { v4 as uuidv4 } from "uuid";
import fs from "fs/promises";
import path from "path";
import { fileURLToPath } from "url";
import { config } from "../config/env.js";
import { logger } from "../utils/logger.js";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const localUploadDir = path.resolve(__dirname, "../../uploads/avatars");

let s3Client = null;

function getClient() {
  if (!s3Client) {
    if (!config.AWS_ACCESS_KEY_ID || !config.AWS_SECRET_ACCESS_KEY) {
      logger.warn("AWS S3 credentials not configured in environment.");
      return null;
    }
    s3Client = new S3Client({
      region: config.AWS_REGION,
      credentials: {
        accessKeyId: config.AWS_ACCESS_KEY_ID,
        secretAccessKey: config.AWS_SECRET_ACCESS_KEY,
      },
    });
  }
  return s3Client;
}

async function saveLocally(fileBuffer, originalFilename) {
  await fs.mkdir(localUploadDir, { recursive: true });
  const ext = path.extname(originalFilename).toLowerCase() || ".png";
  const filename = `${uuidv4().replace(/-/g, "")}${ext}`;
  await fs.writeFile(path.join(localUploadDir, filename), fileBuffer);
  const url = `${config.PUBLIC_BASE_URL.replace(/\/$/, "")}/uploads/avatars/${filename}`;
  logger.info(`Profile image stored locally: ${url}`);
  return { url, key: `avatars/${filename}` };
}

export const s3Service = {
  /**
   * Upload file buffer to S3 (or local uploads/ when AWS is not configured).
   */
  async uploadFile(fileBuffer, originalFilename, contentType) {
    const client = getClient();
    if (!client || !config.AWS_S3_BUCKET) {
      logger.warn("AWS S3 is not configured. Saving profile image locally under /uploads/avatars.");
      return saveLocally(fileBuffer, originalFilename);
    }

    const ext = path.extname(originalFilename).toLowerCase();
    const uniqueKey = `avatars/${uuidv4().replace(/-/g, "")}${ext}`;

    const command = new PutObjectCommand({
      Bucket: config.AWS_S3_BUCKET,
      Key: uniqueKey,
      Body: fileBuffer,
      ContentType: contentType,
    });

    try {
      await client.send(command);
      const url = `https://${config.AWS_S3_BUCKET}.s3.${config.AWS_REGION}.amazonaws.com/${uniqueKey}`;
      logger.info(`File uploaded successfully to S3: ${url}`);
      return { url, key: uniqueKey };
    } catch (err) {
      logger.error(`AWS S3 upload failed: ${err.message}`);
      throw new Error(`AWS S3 upload failed: ${err.message}`);
    }
  },

  /**
   * Extract key from URL and delete from S3.
   */
  async deleteFile(imageUrl) {
    if (!imageUrl) return false;

    if (imageUrl.includes("/uploads/avatars/")) {
      try {
        const filename = imageUrl.split("/uploads/avatars/")[1];
        if (filename) {
          await fs.unlink(path.join(localUploadDir, filename));
          logger.info(`Deleted local avatar: ${filename}`);
        }
      } catch (err) {
        logger.warn(`Local avatar delete skipped: ${err.message}`);
      }
      return true;
    }

    const client = getClient();
    if (!client || !config.AWS_S3_BUCKET) {
      logger.warn("AWS S3 not configured. Skipping S3 deletion.");
      return false;
    }

    try {
      if (!imageUrl.includes("s3") || !imageUrl.includes(".amazonaws.com")) {
        return false;
      }

      let key = "";
      if (imageUrl.includes("avatars/")) {
        key = "avatars/" + imageUrl.split("avatars/")[1];
      } else {
        const parts = imageUrl.split(".amazonaws.com/");
        if (parts.length > 1) {
          key = parts[1];
        } else {
          return false;
        }
      }

      const command = new DeleteObjectCommand({
        Bucket: config.AWS_S3_BUCKET,
        Key: key,
      });

      await client.send(command);
      logger.info(`Successfully deleted S3 object: ${key}`);
      return true;
    } catch (err) {
      logger.error(`Failed to delete S3 object for url ${imageUrl}: ${err.message}`);
      return false;
    }
  },
};
